apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/strategy/ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSet.java

```diff
@@ -95,63 +91,78 @@ protected ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSe
+      NovaTemplateOptions templateOptions = NovaTemplateOptions.class.cast(template.getOptions());
+      String region = template.getLocation().getId();
 
       if (templateOptions.shouldAutoAssignFloatingIp()) {
          checkArgument(novaApi.getFloatingIPApi(region).isPresent(),
                  "Floating IPs are required by options, but the extension is not available! options: %s",
                  templateOptions);
       }
+      if (templateOptions.shouldGenerateKeyPair() || templateOptions.getKeyPairName() != null) {
+         checkArgument(novaApi.getKeyPairApi(region).isPresent(),
+                 "Key Pairs are required by options, but the extension is not available! options: %s", templateOptions);
+      }
 
-      boolean keyPairExtensionPresent = novaApi.getKeyPairApi(region).isPresent();
+      List<Integer> inboundPorts = Ints.asList(templateOptions.getInboundPorts());
+      if (!templateOptions.getGroups().isEmpty() && !inboundPorts.isEmpty()) {
+         checkArgument(novaApi.getSecurityGroupApi(region).isPresent(),
+                 "Security groups are required by options, but the extension is not available! options: %s",
+                 templateOptions);
+      }
+      
+      KeyPair keyPair = null;
       if (templateOptions.shouldGenerateKeyPair()) {
-         checkArgument(keyPairExtensionPresent,
-                  "Key Pairs are required by options, but the extension is not available! options: %s", templateOptions);
-         KeyPair keyPair = keyPairCache.getUnchecked(RegionAndName.fromRegionAndName(region, namingConvention.create()
-                  .sharedNameForGroup(group)));
-         keyPairCache.asMap().put(RegionAndName.fromRegionAndName(region, keyPair.getName()), keyPair);
-         templateOptions.keyPairName(keyPair.getName());
-         tagsBuilder.add(JCLOUDS_KP);
-      } else if (templateOptions.getKeyPairName() != null) {
-         checkArgument(keyPairExtensionPresent,
-                  "Key Pairs are required by options, but the extension is not available! options: %s", templateOptions);
-         if (templateOptions.getLoginPrivateKey() != null) {
-            String pem = templateOptions.getLoginPrivateKey();
-            KeyPair keyPair = KeyPair.builder().name(templateOptions.getKeyPairName())
-                     .fingerprint(fingerprintPrivateKey(pem)).privateKey(pem).build();
-            keyPairCache.asMap().put(RegionAndName.fromRegionAndName(region, keyPair.getName()), keyPair);
+         keyPair = generateKeyPair(region, namingConvention.create().sharedNameForGroup(group));
+         // If a private key has not been explicitly set, configure the auto-generated one
+         if (Strings.isNullOrEmpty(templateOptions.getLoginPrivateKey())) {
+            templateOptions.overrideLoginPrivateKey(keyPair.getPrivateKey());
          }
+      } else if (templateOptions.getKeyPairName() != null && templateOptions.getLoginPrivateKey() != null) {
```

Comment:
Why checking also the private key here? If users configure a key pair we don't care if they also configure the private key.  Creating a node and not accessing it is a valid use case. Perhaps users access it in another connection (where the private key will be configured). I'd remove the private key check from here and always set the keypair when its name is set.